feature: gas meter

vlopes11 opened the pull request:
Blocked by: #792

bkolad commented on examples/demo-nft-module/tests/nft_test.rs:
```diff
@@ -70,7 +76,11 @@ fn transfer() {
     let tmpdir = tempfile::tempdir().unwrap();
-    let mut working_set = WorkingSet::new(ProverStorage::with_path(tmpdir.path()).unwrap());
+    let mut working_set = WorkingSet::new(
+        todo!(),
```
Let's put `gas & gas_price` in a single struct so the `WorkingSet::new` takes less arguments.

bkolad commented on module-system/module-implementations/examples/sov-value-setter/src/lib.rs:
```diff
@@ -30,6 +30,10 @@ pub struct ValueSetter<C: sov_modules_api::Context> {
 
+    /// The gas configuration for the module
+    #[gas]
+    pub gas: DefaultGasConfig,
```
is `[gas]` mandatory? If a module does not do any gas calculation can we skip it?

bkolad commented on module-system/module-implementations/examples/sov-election/src/lib.rs:
```diff
@@ -55,11 +59,16 @@ impl<C: sov_modules_api::Context> sov_modules_api::Module for Election<C> {
+        working_set: &mut WorkingSet<C::Storage, G>,
+    ) -> Result<(), Error>
+    where
+        G: GasUnit,
```
We have to find a more ergonomic way to pass `GasUnit` around. What about putting it in `C: Context` this way `WorkingSet` won't need an explicit `G`

vlopes11 replied:
Very good suggestion; onto that

vlopes11 commented:
Blocked by #938

vlopes11 committed:
feat: add gas meter to working set

This commit introduces `GasMeter`, encapsulated by `WorkingSet`.

It will allow the user to consume scalar gas from the working set, and
define arbitrary price parsed from a constants.json manifest file at
compilation. At each compilation, the `ModuleInfo` derive macro will
parse such file, and set the gas price configuration.

vlopes11 commented on module-system/sov-modules-api/src/gas.rs:
```diff
@@ -0,0 +1,92 @@
+    const ZEROED: Self;
+
+    /// Creates a unit from a multi-dimensional unit with arbitrary dimension.
+    fn from_arbitrary_dimensions(dimensions: &[u64]) -> Self;
```
This function will allow us to convert arbitrary arrays from the JSON file into the concrete, chosen type.

It is designed to have const-like performance, as in the link below
https://rust.godbolt.org/z/rPhaxnPEY

vlopes11 commented on module-system/sov-modules-api/src/lib.rs:
```diff
@@ -305,6 +310,17 @@ pub trait Module {
+    ///
+    /// The scalar gas value will be computed from the price defined on the working set.
+    fn charge_gas(
+        &self,
```
We take `self` to allow the user to implement his own gas config, if he wants to

bkolad replied:
Why do wee need this function in the `Context`? The caller can directly call `working_set.charge_gas(gas)`.

vlopes11 commented on module-system/sov-modules-macros/src/manifest.rs:
```diff
@@ -188,42 +189,29 @@ impl Manifest {
-                #(#fields,)*
-            }
-        };
+        // remove generics, if any
```
Rust will not allow us to declare the type if it is defined via generics (i.e. reduce `Bank<C>` to `Bank`)

vlopes11 committed:
fix lint fmt

vlopes11 committed:
fix ci test expected error string

vlopes11 committed:
Merge branch 'nightly' into vlopes11/feature/gas-meter

bkolad commented on module-system/sov-modules-api/src/default_context.rs:
```diff
@@ -31,6 +31,8 @@ impl Spec for DefaultContext {
 
 #[cfg(feature = "native")]
 impl Context for DefaultContext {
+    type GasUnit = TupleGasUnit<1>;
```
It should be `TupleGasUnit<2>` the same as for the `ZkContext` it is because we need to charge for both executions.

vlopes11 replied:
Makes sense, fixed!

vlopes11 committed:
update default context to 2 dimensions

vlopes11 committed:
Merge branch 'nightly' into vlopes11/feature/gas-meter
